cnd/src/btsieve/ethereum/web3_connector.rs

```diff
@@ -37,15 +31,75 @@ impl LatestBlock for Web3Connector {
     fn latest_block(
         &mut self,
     ) -> Box<dyn Future<Item = Self::Block, Error = anyhow::Error> + Send + 'static> {
-        let web = self.web3.clone();
-        Box::new(
-            web.eth()
-                .block_with_txs(BlockId::Number(BlockNumber::Latest))
-                .from_err(),
-        )
+        let web3 = self.web3.clone();
+        let url = self.url.clone();
+
+        let future = async move {
+            let request = JsonRpcRequest::new("eth_getBlockByNumber", vec![
+                serialize(BlockId::Number(BlockNumber::Latest))?,
+                serialize(true)?,
+            ]);
+
+            let response = web3
+                .post(url)
+                .json(&request)
+                .send()
+                .await?
+                .json::<JsonRpcResponse<crate::ethereum::Block<crate::ethereum::Transaction>>>()
+                .await?;
+
+            let block = match response {
+                JsonRpcResponse::Success { result } => result,
+                JsonRpcResponse::Error { code, message } => {
+                    log::warn!(
+                        "eth_getBlockByNumber request failed with {}: {}",
+                        code,
+                        message
+                    );
+                    return Ok(None);
+                }
+            };
+
+            log::trace!(
+                "Fetched block from web3: {}",
+                block.hash.expect("blocks to have a hash")
```

Comment:
❓ What happens if there was not hash?

Reply:
We panic and the tokio executor thread dies.

I am unaware of a scenario where blocks don't have a hash. In case this is actually legit, it is a serious bug and we should immediately stop IMO. Hence the decision to panic.

Reply:
I expect this to get simpler once we completely remove web3 and have control over those data structures.